### PR TITLE
Add LightGBM tool

### PR DESCRIPTION
**Pull request by simonbray**
This is a tool @khanteymoori and I need for our cheminformatics project.

### CODE REVIEW

File: tools/sklearn/lightgbm.xml [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    </command>
+    <configfiles>
+        <inputs name="inputs"/>
+        <configfile name="lightgbm_script">
+<![CDATA[
+import json
+import lightgbm as lgb
+import numpy as np
+import pandas
+import pickle
```
Comment (qiagu) [inline tools/sklearn/lightgbm.xml]: `pickle` is not used.

Comment (simonbray) [reply tools/sklearn/lightgbm.xml]: Good point, thanks!

**Comment by simonbray** [comment]
@bgruening, can you have a look at this?

There is one test failing, but it's not related to this tool. Not sure why - it completes successfully locally as well as in @qiagu's PR.

File: tools/sklearn/lightgbm.xml [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+import sys
+
+from scipy.io import mmread
+import sklearn.ensemble
```
Comment (qiagu) [inline tools/sklearn/lightgbm.xml]: And `ensemble `

**Comment by qiagu** [comment]
This is great. 
I had an idea to add `lightgbm`, maybe `catboost` as well, to the pipeline tool long time ago. Here we have two options: 1) directly add to the pipeline tool, like we did for `xgboost`; 2) extend this tool to output unfitted pickle object and load it into the pipeline tool via `custom model`.

**Comment by simonbray** [comment]
Hm, I think the disadvantage of 1) is that the tool will need an extra requirement (`lightgbm`) which will not be relevant for most users.

File: tools/sklearn/lightgbm.xml [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+if params["selected_tasks"]["selected_task"] == "train":
+    algorithm = params["selected_tasks"]["selected_algorithms"]["selected_algorithm"]
+    options = params["selected_tasks"]["selected_algorithms"]["options"]
+    if "max_leaf_nodes" in options:
+        options["num_leaves"] = options["max_leaf_nodes"]
+        options.pop("max_leaf_nodes")
+
+    X, y = get_X_y(params, infile1, infile2)
+    lgb_train = lgb.Dataset(X, y)
+    gbm = lgb.train(options, lgb_train)
```
Comment (qiagu) [inline tools/sklearn/lightgbm.xml]: It seems these estimators don't comply with sklearn `fit`, `predict` APIs. Wrapper is needed to make them work in pipeline.

**Comment by qiagu** [comment]
https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/sklearn.py
I'm glad they support sklearn APIs.
Pipeline has advantages over single estimator if preprocessing steps are needed, which happens most time in my practices. It's well discussed that applying precessing on whole dataset before train test split can lead to incorrect evaluation.

**Comment by qiagu** [comment]
https://github.com/bgruening/galaxytools/pull/943/files#diff-cee8a67d2b46be46df04d32f0a71234cL1-L5
The same error can be solved by updating the test data.
@simonbray You're right. Adding to pipeline would lead to the dependencies change. We can leave that later.

**Comment by simonbray** [comment]
@qiagu, what do you think needs to be done before this can be merged? Should I make an option to output a pickle object, or should this be done later?

File: tools/sklearn/lightgbm.xml [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+import lightgbm as lgb
+import numpy as np
+import pandas
+import sys
+
+from scipy.io import mmread
+from galaxy_ml.utils import load_model, get_X_y
+
+
+N_JOBS = int(__import__('os').environ.get('GALAXY_SLOTS', 1))
```
Comment (anuprulez) [inline tools/sklearn/lightgbm.xml]: Do you use this variable (`N_JOBS`) anywhere else in the tool definition?

File: tools/sklearn/lightgbm.xml [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    </tests>
+    <help><![CDATA[
+**What it does**
+
+LightGBM is a gradient boosting framework that uses tree based learning algorithms. It 
+is designed to be distributed and efficient with the following advantages:
+
+* Faster training speed and higher efficiency.
+* Lower memory usage.
+* Better accuracy.
```
Comment (anuprulez) [inline tools/sklearn/lightgbm.xml]: `Faster, better ...` than which algorithm? If you know, please mention it

File: tools/sklearn/lightgbm.xml [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        <inputs name="inputs"/>
+        <configfile name="lightgbm_script">
+<![CDATA[
+import json
+import lightgbm as lgb
+import numpy as np
+import pandas
+import sys
+
+from scipy.io import mmread
```
Comment (anuprulez) [inline tools/sklearn/lightgbm.xml]: do you use `mmread` imported method anywhere else?

File: tools/sklearn/lightgbm.xml [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+from galaxy_ml.utils import load_model, get_X_y
```
Comment (anuprulez) [inline tools/sklearn/lightgbm.xml]: same with `load_model`?

File: tools/sklearn/lightgbm.xml [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    python '$lightgbm_script' '$inputs'
+]]>
+    </command>
+    <configfiles>
+        <inputs name="inputs"/>
+        <configfile name="lightgbm_script">
+<![CDATA[
+import json
+import lightgbm as lgb
+import numpy as np
```
Comment (anuprulez) [inline tools/sklearn/lightgbm.xml]: I do not see `np` being used anywhere? Please check

File: tools/sklearn/lightgbm.xml (hunk 5 above)
Comment (simonbray) [reply tools/sklearn/lightgbm.xml]: You are right, this isn't very objective. I have removed this section. Thanks for the comments!

**Comment by qiagu** [comment]
This is fine for this tool and PR. We can add lightGBM into galaxy-ml utils in next version. Then add the classifier and regressor in pipeline tool and be able to tune hyperparameters

**Pull request merged**